### DISCUSSION TITLE
Control Aciduino via MIDI-IN and simple MIDI-Keyboard

### DISCUSSION THREAD

**Discussion by ErichHeinemann** [copych/AcidBox#14] [Ideas]
Hi, I suggest to add an Control-MIDI-Channel to force SEED-Notes like we know from Arranger-Keyboards or from Arpeggiators. Second, we could add a function to mute/unmute single instruments by one octave. Start / Stop via on key. Force a Break with one key? Then an old 25-Key-MIDI-Keyboard could become the right controller/housing for AcidBox.

**Comment by ErichHeinemann** [comment]
here is some of my code to collect the used MIIDI-Notes comming from an MIDI-Keyboard.... currently it does not effect the actual playing note-Set ... but after a while the last notes are used...  I believe, You could identify the direction... simply a way to control the played noteset via a MIDI-keyboard.

in  inline void handleNoteOn(uint8_t .. it now looks like this, - i added a CTRL_CHAN in config.h

```
  else if (inChannel == SYNTH2_MIDI_CHAN )  {Synth2.on_midi_noteON(inNote, inVelocity);}
  else if (inChannel = CTRL_CHAN ){
     regmidi_note( inNote );
  }

```

 in AcidBanger I added and changed this

```
// register Midi Notes
void regmidi_note( byte note_num ){
  for ( uint8_t i = 3; i > 0; i-- ){

    // Check if note is already registered
    // tdb

    // shift notes like Pete Midi
    my_midi_notes[i]=  (my_midi_notes[i-1]);    
    my_midi_rootnotes[i]= (my_midi_rootnotes[i-1]);
    
    DEBF("CTRL-Note-Statistics %d %d \r\n", my_midi_notes[i], i);  
  }  
  my_midi_notes[0]= note_num;    
    DEBF("CTRL-Note-Statistics %d %d \r\n", my_midi_notes[0], 0);  
  my_midi_rootnotes[0]= note_num%12 + 12; // don´t use the lowest Octave ... perehaps +12 or +24 is good.    
}


static byte generate_note_set(uint8_t *note_set) {

  // Random root note
  byte root = myRandom(15) + 28;

  if( my_midi_notes[0] != -1 ){
     root =  my_midi_notes[0];
  }

  // Random note set
  byte set = myRandom(ARRAY_SIZE(offset_choices));

  // Copy notes from note set and offset them by "root note"
  int i;
  for (i = 0; i < MaxNoteSet; i++) {
    int8_t note = offset_choices[set][i];
    if (note < 0)
      break;
    if( i < 4 &&  my_midi_notes[i] != -1 ){
      *note_set++ = my_midi_notes[i]; // so I only copy the first 4 steps into the set ... 
    }else{
      *note_set++ = root + note;
    }  
  }
  
  return i;
}

```

**Comment by copych** [comment]
Yes, thanks for the suggestion and the code!
It is definitely a great feature for performing live =) I'll check how we can make it efficient and easy-to-use!

**Reply by copych** [reply]
@ErichHeinemann what are your thoughts on the following implementation?
We read pressed midi notes at the beginning of each bar, and if there are some, we store a new notes set, assuming the lowest note to be the root, then generate a new pattern based on this set and make it active.
Also we need to decide what should be the performance mode in general: which controls should be fully-automated, which semi-automated, and which should stay manual? I mean randomizing patterns, breaks, "auto-turning knobs" and so on.

**Comment by ErichHeinemann** [comment]
Hi CopyCH,
that sounds good.
As the user is using a keyboard, we could use a note to switch between randomised patterns and performance-mode.
The next key beside this key could be used to enable/disable the ramps as „Automation"

If someone is using an old 25key MIDI-Keyboard, the user probably could reserve some keys for this function. These special keys then shouldn’t trigger notes.

Could this be an acceptable way?

Regards
Erich

**Reply by copych** [reply]
Actually, there can be 2 general setups: midi assisted and stand-alone. In case of stand-alone we can make some kind of all-purpose all-weather box, including a PCB design and so on. But when it concerns midi, every other person has a very unique set of equipment. Some has a reach 88 weighted keys, some just 16 keys without velocity... So we could make control midi options configurable.
BTW, do you think that forced 4-bar song structure is a good solution for performing live?

**Reply by copych** [reply]
Even further, this midi control config should be stored in flash memory and be editable in run-time (though LCD seems to be an unneeded extra)

**Comment by ErichHeinemann** [comment]
Hi CopyCH,

this evening I did some tests… this is not all well done, but partly works.

MIDI-Keys-Control is partly implemented in MIDI-Handlers… but is not yet tested.

MIDI-Clock und Start/Stop via MIDI works, .. perhaps the Clock-Devider is wrong. I only tested only the basic function with a TD3 and 2 headphones on top of my head.
External Start/Stop toggles between internal and external clock.
I had connected 3 buttons to GPIO32 for Start/Stop and a „Do a Break“ on GPIO33.

The analog MUX works. perhaps we could do it like in other hybrid synths, -  the Pot-value works directly or changes only the CC_Values after it crossed the old stored value 

My TEST-Code is there... I am not good in using Github ... so I made a Copy for testing and sharing.
Next step could be more Pots, and Display…


https://github.com/ErichHeinemann/AcidBox-other_coresetup￼
ErichHeinemann/AcidBox-other_coresetup: ESP32 synth: all-in-one acid combo TB303 + TB303 + drum machine + fx chain, cd-quality, no lag
github.com

**Reply by copych** [reply]
Hi @ErichHeinemann , I am starting to implement playback control functions in order to bind them to the choosen controller later. And the more I think on the matter the more I can conclude that having an RGB LED strip over the midi keyboard to indicate current state and have some kind of hints during performance would be a great addition! I only doubt if it would be energy-wise, because to lit all of the WS2812b LEDS at max it would take 25*0.06A = 1.5A @ 5V.

**Comment by ErichHeinemann** [comment]
most LEDs work with 3.3V too …

If not, then run the first with 4.3 Volts as VCC and it will accept 3.3 on Data In!!
From this WS9212-LED use its Data-Out to the second LEDs Data-In and power this second LED with 5 Volts…. 

Works perfekt.
use a simple diode 4001 or something else to reduce to Voltage from 5 Volts to 4.3 Volts. :)
That workaround works perfekt.

I am using WS8212-Leds in another project and they work with 3.3 V as VCC perfectly.

Regards
Erich

**Comment by copych** [comment]
@ErichHeinemann , I have tested a 5V 60 LEDs strip yesterday, and it worked well. Also, I have found my old good MAudio keystation32 mini,  and 60Leds/meter strip fits well along with the white keys. The challenge for me now is how to (if possible at all) connect the keyboard as a slave to my ESP32s3 running USB OTG as a host. What kind of keyboard do you have? Key size, 5-pin MIDI or USB?

**Reply by ErichHeinemann** [reply]
Hi, I only use old MIDI .. Din-plugs.. I would use an old MIDI-Keyboard with 25 keys for this.
